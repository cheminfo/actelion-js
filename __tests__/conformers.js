'use strict';

const OCL = require('../core');

const { Molecule, ConformerGenerator } = OCL;

describe('ConformerGenerator', () => {
  it('should return one conformer', () => {
    const mol = Molecule.fromSmiles('COCCON');

    const gen = new ConformerGenerator(1);
    expect(gen.getConformerCount()).toBe(0);

    expect(mol.getAtomZ(0)).toBe(0);
    const result = gen.getOneConformerAsMolecule(mol);
    expect(gen.getConformerCount()).toBe(1);
    expect(result).toBe(mol);
    expect(result.getAtomZ(0)).not.toBe(0);
  });

  it('should generate conformers', () => {
    const mol = Molecule.fromSmiles('COCCON');

    const gen = new ConformerGenerator(1);
    expect(gen.getPotentialConformerCount()).toBe(1);
    expect(gen.getConformerCount()).toBe(0);
    expect(gen.getPreviousConformerContribution()).toBe(1);
    gen.initializeConformers(mol);
    expect(gen.getPotentialConformerCount()).toBeGreaterThan(1);

    const conf1 = gen.getNextConformerAsMolecule();
    expect(gen.getConformerCount()).toBe(1);
    expect(gen.getPreviousConformerContribution()).toBeCloseTo(0.155, 2);
    const conf1Molfile = conf1.toMolfile();
    expect(conf1Molfile).toMatchSnapshot();

    const conf2 = gen.getNextConformerAsMolecule();
    expect(gen.getConformerCount()).toBe(2);
    const conf2Molfile = conf2.toMolfile();
    expect(conf2Molfile).toMatchSnapshot();

    expect(conf2).not.toBe(conf1);
    expect(conf2Molfile).not.toBe(conf1Molfile);
  });

  it('should be random if initialized with 0', () => {
    const mol = Molecule.fromSmiles('COCCON');

    const gen1 = new ConformerGenerator(0);
    expect(gen1.getConformerCount()).toBe(0);
    gen1.initializeConformers(mol);
    expect(gen1.getConformerCount()).toBe(0);
    const conf1 = gen1.getNextConformerAsMolecule().toMolfile();
    expect(gen1.getConformerCount()).toBe(1);

    const gen2 = new ConformerGenerator(0);
    expect(gen2.getConformerCount()).toBe(0);
    gen2.initializeConformers(mol);
    expect(gen2.getConformerCount()).toBe(0);
    const conf2 = gen2.getNextConformerAsMolecule().toMolfile();
    expect(gen2.getConformerCount()).toBe(1);

    expect(conf2).not.toBe(conf1);
  });

  it('should use and return the passed molecule', () => {
    const mol = Molecule.fromSmiles('COCCON');

    const gen = new ConformerGenerator(1);
    gen.initializeConformers(mol);

    expect(mol.getAtomZ(0)).toBe(0);

    const conf = gen.getNextConformerAsMolecule(mol);
    expect(conf).toBe(mol);
    expect(mol.getAtomZ(0)).not.toBe(0);
  });
});
