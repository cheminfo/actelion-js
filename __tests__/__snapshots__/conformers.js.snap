// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConformerGenerator should generate conformers 1`] = `
"
Actelion Java MolfileCreator 1.0

 15 14  0  0  0  0  0  0  0  0999 V2000
   -1.3715   -1.7858    5.1340 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.3608   -1.1727    4.4711 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.1032    0.0932    4.9457 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.2275    0.6140    4.0808 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.7501    0.9730    2.7964 O   0  0  0  0  0  0  0  0  0  0  0  0
    1.6887    1.4865    1.8502 N   0  0  0  0  0  0  0  0  0  0  0  0
   -1.7604   -1.3040    5.8725 H   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1166   -1.9921    4.4969 H   0  0  0  0  0  0  0  0  0  0  0  0
   -1.0700   -2.6696    5.4976 H   0  0  0  0  0  0  0  0  0  0  0  0
   -0.6316    0.7507    4.9839 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.4208   -0.0415    5.8810 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.6997    1.3387    4.5418 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.9202   -0.1290    3.9610 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.4246    1.8306    2.1860 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.8525    0.9191    1.2337 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  1  0  0  0  0
  4  5  1  0  0  0  0
  5  6  1  0  0  0  0
  1  7  1  0  0  0  0
  1  8  1  0  0  0  0
  1  9  1  0  0  0  0
  3 10  1  0  0  0  0
  3 11  1  0  0  0  0
  4 12  1  0  0  0  0
  4 13  1  0  0  0  0
  6 14  1  0  0  0  0
  6 15  1  0  0  0  0
M  END
"
`;

exports[`ConformerGenerator should generate conformers 2`] = `
"
Actelion Java MolfileCreator 1.0

 15 14  0  0  0  0  0  0  0  0999 V2000
    0.2977   -2.2932    4.8553 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.3608   -1.1727    4.4711 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.1032    0.0932    4.9457 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.2275    0.6140    4.0808 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.6863    1.8685    4.5523 O   0  0  0  0  0  0  0  0  0  0  0  0
    2.4759    1.8909    5.7422 N   0  0  0  0  0  0  0  0  0  0  0  0
    1.0255   -2.1767    5.4763 H   0  0  0  0  0  0  0  0  0  0  0  0
   -0.3455   -2.9343    5.2782 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.6760   -2.7705    4.0594 H   0  0  0  0  0  0  0  0  0  0  0  0
   -0.6316    0.7507    4.9839 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.4208   -0.0415    5.8810 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.9347   -0.0604    4.0047 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.8750    0.7468    3.1297 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.3892    1.1970    6.2748 H   0  0  0  0  0  0  0  0  0  0  0  0
    3.2944    2.0627    5.5703 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  1  0  0  0  0
  4  5  1  0  0  0  0
  5  6  1  0  0  0  0
  1  7  1  0  0  0  0
  1  8  1  0  0  0  0
  1  9  1  0  0  0  0
  3 10  1  0  0  0  0
  3 11  1  0  0  0  0
  4 12  1  0  0  0  0
  4 13  1  0  0  0  0
  6 14  1  0  0  0  0
  6 15  1  0  0  0  0
M  END
"
`;
